script/benchmark.js

```diff
@@ -0,0 +1,122 @@
+'use strict';
+
+const { performance, PerformanceObserver } = require('perf_hooks');
+const { spawnSync, spawn } = require('child_process');
+const os = require('os');
+const fs = require('hexo-fs');
+const path = require('path');
+let npmScript = 'npm';
+let npxScript = 'npx';
+const testDir = path.resolve('.tmp-hexo-theme-unit-test');
+
+if (os.platform() === 'win32') {
+  npmScript = 'npm.cmd';
+  npxScript = 'npx.cmd';
+}
+
+(async () => {
+  init();
+  // make sure no cache files
+  cleanup();
+  await run_benchmark('Cold processing');
+  await run_benchmark('Hot processing');
+  cleanup();
+  await run_benchmark('Another Cold processing');
+  cleanup();
+})();
+
+async function run_benchmark(name) {
+  return new Promise(resolve => {
+    const result = {};
+    const obs = new PerformanceObserver(list => {
+      const { name, duration } = list.getEntries()[0];
+      result[name] = {
+        'Cost time (s)': `${(duration / 1000).toFixed(2)}s`
+      };
+    });
+    obs.observe({ entryTypes: ['measure'] });
+
+    const npx = spawn(npxScript, ['--no-install', 'hexo', 'g', '--debug'], { cwd: testDir });
+    npx.stdout.on('data', data => {
+      const string = data.toString('utf-8');
+      if (/Hexo version/.test(string)) {
+        performance.mark('hexo-begin');
+      } else if (/Start processing/.test(string)) {
+        performance.mark('processing');
+      } else if (/Files loaded/.test(string)) {
+        performance.mark('file-loaded');
+      } else if (/generated in/.test(string)) {
+        performance.mark('generated');
+      } else if (/Database saved/.test(string)) {
+        performance.mark('database-saved');
+      }
+    });
+
+    npx.on('close', () => {
+      performance.measure('Load Plugin/Scripts/Database', 'hexo-begin', 'processing');
+      performance.measure('Process Source', 'processing', 'file-loaded');
+      performance.measure('Render Files', 'file-loaded', 'generated');
+      performance.measure('Save Database', 'generated', 'database-saved');
+      performance.measure('Total time', 'hexo-begin', 'database-saved');
+      obs.disconnect();
+      resolve();
+    });
+
+    process.once('exit', () => {
+      console.log(name);
+      console.table(result);
```

Comment:
https://github.com/hexojs/hexo/blob/75201130fef4a02f27e239200714d5815f97ee3a/test/benchmark.sh#L39-L44

Performance regression detection could be added (simply a threshold of `20s` will be fine).

`chalk` could be utilized for colorize output.